### PR TITLE
MISUV-6363 -  SCP001| Cease Journey Check Details Page Controller Unification/Refactor

### PR DESCRIPTION
**Pull request by JordHemsworth**

### CODE REVIEW

File: app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            }
+        }
+
+      case (Right(None), Right(Some(cessationEndDate))) =>
+        val propertyIncomeSources = (incomeSourceType) match {
+          case (UkProperty) => user.incomeSources.properties.find(x => x.isUkProperty && !x.isCeased)
+          case (ForeignProperty) => user.incomeSources.properties.find(x => x.isForeignProperty && !x.isCeased)
+          case (SelfEmployment) => None
+        }
+        if(propertyIncomeSources.isEmpty){
```
Comment (EdwardDennis) [inline app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala:180]: Should fix:
Can you format this file? 🙏

File: app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+          implicit mtdItUser =>
+            handleRequest(
+              sources = mtdItUser.incomeSources,
+              isAgent = true,
+              incomeSourceType = incomeSourceType
+            )
+        }
+  }
+
+  def handleSubmitRequest(isAgent: Boolean, incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_], request: Request[_]): Future[Result] = withIncomeSourcesFS {
```
Comment (EdwardDennis) [inline app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala:146]: Could fix:
This function could be split into smaller ones for readability

File: app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala (hunk 1 above)
Comment (JordHemsworth) [reply app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala]: I've just pushed auto format, lmk it needs further

File: app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    for {
+      incomeSourceId <- incomeSourceIdFuture
+      cessationEndDate <- cessationEndDateFuture
+    } yield (incomeSourceId, cessationEndDate)
+  }
+
+  def handleRequest(sources: IncomeSourceDetailsModel, isAgent: Boolean, origin: Option[String] = None, incomeSourceType: IncomeSourceType)
+                   (implicit user: MtdItUser[_], hc: HeaderCarrier, request: Request[_]): Future[Result] = withIncomeSourcesFS {
```
Comment (EdwardDennis) [inline app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala:70]: Should fix: I don't think hc or request is being used

File: app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala (hunk 2 above)
Comment (EdwardDennis) [reply app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala:146]: Maybe one submit function which calls other functions as and when required e.g. an updateCessationDate function?
I've just realised the cyclomatic complexity of this function is 28 which is very high, it'd be good if you could reduce this somehow

File: app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    } else {
+        controllers.incomeSources.cease.routes.CeaseCheckIncomeSourceDetailsController.submit(SelfEmployment)
+    }
+}
+
+
+@getPropertyChangeHref = @{
+    if(viewModel.incomeSourceType.toString == "UkProperty") {
+        if(isAgent){
+        controllers.incomeSources.cease.routes.IncomeSourceEndDateController.showChangeAgent(None, UkProperty).url
```
Comment (EdwardDennis) [inline app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html:70]: needs a tab

File: app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        } else {
+        controllers.incomeSources.cease.routes.IncomeSourceEndDateController.showChange(None, UkProperty).url
```
Comment (EdwardDennis) [inline app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html:72]: this too

File: app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        }
+    } else {
+        if(isAgent){
+        controllers.incomeSources.cease.routes.IncomeSourceEndDateController.showChangeAgent(None, ForeignProperty).url
```
Comment (EdwardDennis) [inline app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html:76]: tab

File: app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        } else {
+        controllers.incomeSources.cease.routes.IncomeSourceEndDateController.showChange(None, ForeignProperty).url
```
Comment (EdwardDennis) [inline app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html:78]: tab

File: app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                        ))
+                    ))
+                ),
+                SummaryListRow(
+                    key = Key(Text(getMessage("businessName"))),
+                    value = Value(Text(
+                        if(viewModel.tradingName){
+                            viewModel.tradingName.get
+                        } else {
+                        messages("incomeSources.generic.unknown")
```
Comment (EdwardDennis) [inline app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html:150]: this line needs a tab

File: app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+                        }
+                    ))
+                ),
+                SummaryListRow(
+                    key = Key(Text(getMessage("businessAddress"))),
+                    value = Value(HtmlContent(
+                        if(viewModel.address) {
+                            getLongAddressFromBusinessAddressDetails(viewModel.address.get).toString
+                        } else {
+                        messages("incomeSources.generic.unknown")
```
Comment (EdwardDennis) [inline app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html:160]: tab

File: it/connectors/NavBarEnumFsConnectorISpec.scala [hunk 10]
```diff
@@ -1,60 +1,60 @@
...
-        await(result) shouldBe None
-
-        BtaNavBarPartialConnectorStub.verifyNavlinksContent(1)
-
-      }
-    }
-  }
-
-}
+//
```
Comment (EdwardDennis) [inline it/connectors/NavBarEnumFsConnectorISpec.scala:1]: Commited by mistake

File: it/controllers/agent/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+          elementAttributeBySelector("form", "action")(formActionFP)
+        )
+      }
+    }
+  }
+
+  s"calling POST ${showCheckCeaseForeignPropertyDetailsControllerUrl}" should {
+    s"redirect to $redirectUriFP" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with UK property")
```
Comment (EdwardDennis) [inline it/controllers/agent/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala:194]: I think this should be foreign property

File: it/controllers/agent/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+          elementAttributeBySelector("form", "action")(formActionSE)
+        )
+      }
+    }
+  }
+
+  s"calling POST ${showCheckCeaseBusinessDetailsControllerUrl}" should {
+    s"redirect to $redirectUriSE" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with multiple business and property")
```
Comment (EdwardDennis) [inline it/controllers/agent/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala:99]: I think you're only stubbing a single business (line 102)

File: it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+  val requestFP: UpdateIncomeSourceRequestModel = UpdateIncomeSourceRequestModel(
+    nino = testNino,
+    incomeSourceId = foreignProperty.incomeSourceId,
+    cessation = Some(Cessation(true, Some(LocalDate.parse(cessationDate))))
+  )
+
+  s"calling GET ${showCheckCeaseBusinessDetailsControllerUrl}" should {
+    "render the Cease Business Details Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with multiple business and property")
```
Comment (EdwardDennis) [inline it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala:68]: This is only a single business response as well

File: it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+          elementTextBySelector("dl:nth-of-type(1) div:nth-of-type(2) dt:nth-of-type(1)")(messagesAPI("incomeSources.ceaseBusiness.checkDetails.businessName")),
+          elementTextBySelector("dl:nth-of-type(1) div:nth-of-type(2) dd:nth-of-type(1)")(testBusinessName),
+          elementAttributeBySelector("form", "action")(formActionSE)
+        )
+      }
+    }
+
+    "render the Cease Business Page with unknown address and title" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with multiple business and property")
```
Comment (EdwardDennis) [inline it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala:91]: Same here, it's a single business response (but you might want to add that it's not got an address)

File: it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+          redirectURI(redirectUriSE),
+        )
+      }
+    }
+  }
+
+  s"calling GET ${showCheckCeaseUKPropertyDetailsControllerUrl}" should {
+    "render the Cease UK Property Details Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with multiple business and property")
```
Comment (EdwardDennis) [inline it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala:140]: Should be UK property

File: it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+          redirectURI(redirectUriUK),
+        )
+      }
+    }
+  }
+
+  s"calling GET ${showCheckCeaseForeignPropertyDetailsControllerUrl}" should {
+    "render the Cease Foreign Property Details Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with multiple business and property")
```
Comment (EdwardDennis) [inline it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala:185]: Should be Foreign property

File: it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+          elementAttributeBySelector("form", "action")(formActionFP)
+        )
+      }
+    }
+  }
+
+  s"calling POST ${showCheckCeaseForeignPropertyDetailsControllerUrl}" should {
+    s"redirect to $redirectUriFP" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with multiple business and property")
```
Comment (EdwardDennis) [inline it/controllers/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala:207]: FP also

File: app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+      case ForeignProperty => sessionService.get(ceaseForeignPropertyEndDate)
+    }
+
+    for {
+      incomeSourceId <- incomeSourceIdFuture
+      cessationEndDate <- cessationEndDateFuture
+    } yield (incomeSourceId, cessationEndDate)
+  }
+
+  def handleRequest(sources: IncomeSourceDetailsModel, isAgent: Boolean, origin: Option[String] = None, incomeSourceType: IncomeSourceType)
```
[the remark this answers is not in the document]
Comment (JordHemsworth) [reply app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala]: fixed in lastest push

File: app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html (hunk 4 above)
Comment (JordHemsworth) [reply app/views/incomeSources/cease/CeaseCheckIncomeSourceDetails.scala.html]: view formatted

File: it/connectors/NavBarEnumFsConnectorISpec.scala (hunk 10 above)
Comment (JordHemsworth) [reply it/connectors/NavBarEnumFsConnectorISpec.scala]: fixed

File: it/controllers/agent/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala (hunk 12 above)
Comment (JordHemsworth) [reply it/controllers/agent/incomeSources/cease/CheckCeaseIncomeSourceDetailsControllerISpec.scala]: addressed all test comments

File: app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala (hunk 2 above)
Comment (JordHemsworth) [reply app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala:145]: WIP could pick up when you're free

**Review by EdwardDennis — APPROVED** [review]

**Pull request merged**